server/src/main/java/org/elasticsearch/cluster/ClusterStateObserver.java

```diff
@@ -273,20 +274,32 @@ public String toString() {
     /**
      * The observer considers two cluster states to be the same if they have the same version and master node id (i.e. null or set)
      */
-    private static class StoredState {
-        private final String masterNodeId;
+    public static class StoredState {
+
+        @Nullable
+        private final DiscoveryNode masterNode;
         private final long version;
 
-        StoredState(ClusterState clusterState) {
-            this.masterNodeId = clusterState.nodes().getMasterNodeId();
+        public StoredState(ClusterState clusterState) {
+            this.masterNode = clusterState.nodes().getMasterNode();
```

Comment:
Fair point. I went for this mainly because I got surprised by the fact that the `MasterNodeChangePredicate` does a different check from the `isOlderOrDifferentMaster` method. It seems that both methods should be equally applicable and it might be nice to just align this somehow. I'd have to pass around the `DiscoveryNode` for the master `TransportMasterNode` (in a version that doesn't adjust `ClusterStateObserver.StoredState`) anyway or pass two id strings to keep the current logic.  Seemed nicer to just cabinalize this thing to not have any behaviour change for sure?
Maybe move `StoredState` out of `ClusterStateObserver` into a separate class and call it like `UniqueState` or something to use for this kind of tracking of the current coordinates of the CS?

Reply:
`MasterNodeChangePredicate#build` dates back well before 7.0 but these days the cluster state version persists across restarts so it should be fine to just check the master node's persistent ID and then fall through to `newState.version() > currentVersion`. In fact, isn't it sufficient just to check the version?

Reply:
> In fact, isn't it sufficient just to check the version?

🤦 ++  ... In fact it's all these spots, not just the CS Observer directly but also the predicate building. They all just need to wait for another CS at a newer version to be published to have the same mechanics they have now I think.

-> pushed https://github.com/elastic/elasticsearch/pull/90724/commits/b46c772daf7d095d308c9876504723ae640b40a7 to way simplify the whole thing, thanks!